internal/client/aws.go

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 Cloudbase Solutions SRL
+//
+//    Licensed under the Apache License, Version 2.0 (the "License"); you may
+//    not use this file except in compliance with the License. You may obtain
+//    a copy of the License at
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+//    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+//    License for the specific language governing permissions and limitations
+//    under the License.
+
+package client
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/cloudbase/garm-provider-aws/config"
+	"github.com/cloudbase/garm-provider-aws/internal/spec"
+
+	"github.com/cloudbase/garm-provider-common/errors"
+)
+
+func NewAwsCli(ctx context.Context, cfg *config.Config) (*AwsCli, error) {
+	cliCfg, err := cfg.GetAWSConfig(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get AWS cli context: %w", err)
+	}
+
+	client := ec2.NewFromConfig(cliCfg)
+	awsCli := &AwsCli{
+		cfg:    cfg,
+		client: *client,
+		region: cfg.Region,
+	}
+
+	return awsCli, nil
+}
+
+type AwsCli struct {
+	cfg *config.Config
+
+	client ec2.Client
+	region string
+}
+
+func (a *AwsCli) StartInstance(ctx context.Context, vmName string) error {
+	_, err := a.client.StartInstances(ctx, &ec2.StartInstancesInput{
+		InstanceIds: []string{vmName},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to start instance: %w", err)
+	}
+
+	return nil
+}
+
+func (a *AwsCli) StopInstance(ctx context.Context, vmName string) error {
+	_, err := a.client.StopInstances(ctx, &ec2.StopInstancesInput{
+		InstanceIds: []string{vmName},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to stop instance: %w", err)
+	}
+
+	return nil
+}
+
+func (a *AwsCli) FindInstances(ctx context.Context, controllerID, instanceName string) (*[]types.Instance, error) {
```

Comment:
Why return a pointer to an array? Feel free to return `[]types.Instance{}` if that's what `DescribeInstances()` returns.